### PR TITLE
GH-9846: Migrate to idiomatic error handling in model/client4.go

### LINKED ISSUE #9846: [Help Wanted] [MM-13070] Migrate to idiomatic error handling in the file `model/client4.go` in the mattermost-server repo.
If you're interested please comment here and come [join our "Contributors" community channel](https://pre-release.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://pre-release.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-13070)

This ticket is best suited to someone who is already familiar with what the "idiomatic error handling" migration is about. You can discuss this with us on the Mattermost server linked above, and see examples of these changes at commits such as: [https://github.com/mattermost/mattermost-server/commit/a6fa2b72d1375e554cc6f2e5ca2fd4884a8509f4]

Although `golint` suggests changing `fooId` to `fooID`, please don’t make these changes as part of this PR, as we have decided to keep the codebase consistent the way it is for the time being. At a future date, we will consider migrating the whole codebase.

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
Migrate to idiomatic error handling in the file `model/client4.go` in the mattermost-server repo.

#### Ticket Link
Fixes #9846

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] All new/modified APIs include changes to the drivers
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: model/client4.go [hunk 1]
```diff
@@ -3041,147 +3073,147 @@ func (c *Client4) UploadBrandImage(data []byte) (bool, *Response) {
...
 		return nil, BuildErrorResponse(r, err)
-	} else {
-		defer closeBody(r)
-		return OAuthAppFromJson(r.Body), BuildResponse(r)
 	}
+	defer closeBody(r)
+	return OAuthAppFromJson(r.Body), BuildResponse(r)
 }
 
 // UpdateOAuthApp
```
Comment (hanzei) [inline model/client4.go]: Documentation for this methods is missing. Should we add some in this PR?

File: model/client4.go [hunk 2]
```diff
@@ -3735,161 +3774,157 @@ func (c *Client4) UploadPlugin(file io.Reader) (*Manifest, *Response) {
...
 		return "", BuildErrorResponse(r, err)
-	} else {
-		defer closeBody(r)
-		return MapFromJson(r.Body)["location"], BuildResponse(r)
 	}
+	defer closeBody(r)
+	return MapFromJson(r.Body)["location"], BuildResponse(r)
 }
 
 func (c *Client4) RegisteTermsOfServiceAction(userId, termsOfServiceId string, accepted bool) (*bool, *Response) {
```
Comment (hanzei) [inline model/client4.go]: Documentation for this methods is missing. Should we add some in this PR?

File: model/client4.go [hunk 3]
```diff
@@ -2847,175 +2873,178 @@ func (c *Client4) UploadSamlPrivateCertificate(data []byte, filename string) (bo
...
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, BuildErrorResponse(r, NewAppError("GetBrandImage", "model.client.read_file.app_error", nil, err.Error(), r.StatusCode))
+	}
+
+	return data, BuildResponse(r)
 }
 
 func (c *Client4) DeleteBrandImage() *Response {
```
Comment (hanzei) [inline model/client4.go]: Documentation for this methods is missing. Should we add some in this PR?

File: model/client4.go [hunk 4]
```diff
@@ -2280,256 +2300,259 @@ func (c *Client4) UploadFileAsRequestBody(data []byte, channelId string, filenam
...
 }
 
 // General/System Section
 
 // GetPing will return ok if the running goRoutines are below the threshold and unhealthy for above.
 func (c *Client4) GetPing() (string, *Response) {
-	if r, err := c.DoApiGet(c.GetSystemRoute()+"/ping", ""); r != nil && r.StatusCode == 500 {
+	r, err := c.DoApiGet(c.GetSystemRoute()+"/ping", "")
+	if r != nil && r.StatusCode == 500 {
 		defer r.Body.Close()
```
Comment (hanzei) [inline model/client4.go]: I'm not sure why this code line is here. I don't see a need to explicitly close the request body. But 0/5.

File: model/client4.go [hunk 5]
```diff
@@ -1764,382 +1779,380 @@ func (c *Client4) GetChannelMembersTimezones(channelId string) ([]string, *Respo
...
 		return nil, BuildErrorResponse(r, err)
-	} else {
-		defer closeBody(r)
-		return ChannelFromJson(r.Body), BuildResponse(r)
 	}
+	defer closeBody(r)
+	return ChannelFromJson(r.Body), BuildResponse(r)
 }
 
 func (c *Client4) GetChannelByNameForTeamNameIncludeDeleted(channelName, teamName string, etag string) (*Channel, *Response) {
```
Comment (hanzei) [inline model/client4.go]: Documentation for this methods is missing. Should we add some in this PR?

File: model/client4.go [hunk 6]
```diff
@@ -1764,382 +1779,380 @@ func (c *Client4) GetChannelMembersTimezones(channelId string) ([]string, *Respo
...
 		return nil, BuildErrorResponse(r, err)
-	} else {
-		defer closeBody(r)
-		return ChannelFromJson(r.Body), BuildResponse(r)
 	}
+	defer closeBody(r)
+	return ChannelFromJson(r.Body), BuildResponse(r)
 }
 
 func (c *Client4) GetChannelByNameIncludeDeleted(channelName, teamId string, etag string) (*Channel, *Response) {
```
Comment (hanzei) [inline model/client4.go]: Documentation for this methods is missing. Should we add some in this PR?

File: model/client4.go [hunk 7]
```diff
@@ -2282,15 +2296,22 @@ func (c *Client4) UploadFile(data []byte, channelId string, filename string) (*F
...
 
-	if part, err := writer.CreateFormFile("files", filename); err != nil {
+	part, err := writer.CreateFormFile("files", filename)
+	if err != nil {
 		return nil, &Response{Error: NewAppError("UploadPostAttachment", "model.client.upload_post_attachment.file.app_error", nil, err.Error(), http.StatusBadRequest)}
-	} else if _, err = io.Copy(part, bytes.NewBuffer(data)); err != nil {
+	}
+
+	_, err = io.Copy(part, bytes.NewBuffer(data))
+	if err != nil {
```
Comment (grundleborg) [inline model/client4.go]: Not blocking at all, but I notice this one doesn't use the return value other than `err`, so could stay in the `if` if you want to be a perfectionist :)

**Comment by hanzei** [comment]
Added documentation for some methods. Please review @jespino @grundleborg

**Comment by jespino** [comment]
Closing and reopening it to rerun the tests

**Comment by hanzei** [comment]
Rebased in the hopes it fixes the problem...

**Pull request merged**